xarray/core/dataarray.py

```diff
@@ -816,6 +816,118 @@ def swap_dims(self, dims_dict):
         ds = self._to_temp_dataset().swap_dims(dims_dict)
         return self._from_temp_dataset(ds)
 
+    def set_index(self, indexers=None, append=False, inplace=False,
+                  **kw_indexers):
+        """Set DataArray (multi-)indexes using one or more existing coordinates.
+
+        Parameters
+        ----------
+        indexers : dict, optional
+            Dictionary with keys given by dimension names and values given by
+            (lists of) the names of existing coordinates.
+            Any list of multiple names given for a dimension will result as
+            a MultiIndex for that dimension.
+        append : bool, optional
+            If True, append the supplied indexers to the existing indexes.
+            Otherwise replace the existing indexes (default).
+        inplace : bool, optional
+            If True, set new indexes in-place. Otherwise, return a new DataArray
+            object.
+        **kw_indexers : optional
+            Keyword arguments in the same form as ``indexers``.
+
+        Returns
+        -------
+        reindexed : DataArray
+            Another dataarray, with this dataarray's data but replaced coordinates.
+
+        See Also
+        --------
+        DataArray.reset_index
+        """
+        indexers = utils.combine_pos_and_kw_args(indexers, kw_indexers,
+                                                 'set_index')
+        coords, _ = merge_indexes(indexers, self._coords, set(), append=append)
+        if inplace:
+            self._coords = coords
+        else:
+            return self._replace(coords=coords)
+
+    def reset_index(self, dim_levels=None, drop=False, inplace=False,
```

Comment:
Maybe switch the signature to use separate arguments like `reset_index(dim, levels=None, ...)` instead of using a dict/`**kwargs`. This would make the usual use a clearer, e.g., `array.reset_index('record')` instead of `array.reset_index(record=None)`.

Also, after #1017 (optional indexes), the ability to write `array.reset_index('x', drop=True)` for clearing an index could be nice to have.